### ISSUE TITLE
automatic merging guidelines

### ISSUE THREAD

**Issue by StefanKarpinski** [JuliaRegistries/Registrator.jl#204]
These guidelines are intended not as requirements for packages but as very conservative guidelines, which, if your new package or new version of a package meets them, it may be automatically merged.

### New package

1. Normal capitalization
    - name should match `/^[A-Z]\w*[a-z][0-9]?$/`
        - i.e. starts with a capital letter, ASCII alphanumerics only, ends in lowercase
2. Not too short
    - at least five letters
        - you can register names shorter than this, but doing so requires someone to approve
3. Standard initial version number
    - one of `0.0.1`, `0.1.0`, `1.0.0`
4. Repo URL ends with `/$name.jl.git` where `name` is the package name

### New version

1. Sequential version number
    - if the last version was `1.2.3` then the next can be `1.2.4`, `1.3.0` or `2.0.0`
2. Compat for all dependencies
    - all `[deps]` should also have `[compat]` entries (and Julia itself)
    - `[compat]` entries should have upper bounds
3. Version can be installed
    - given the proposed changes to the registry, can we resolve and install the new version of the package?
4. Version can be loaded
    - once it's been installed (and built?), can we load the code?

**Comment by ararslan** [comment]
> ends in lowercase

? Many packages do not follow this, e.g. HTTP.

**Comment by StefanKarpinski** [comment]
> These guidelines are intended not as requirements for packages but as very conservative guidelines, which, if your new package or new version of a package meets them, it may be automatically merged.

**Comment by ararslan** [comment]
Is that to catch non-obvious acronyms or something?

**Comment by StefanKarpinski** [comment]
Yes, or people who just capitalize an entire package name for no reason, e.g. MYPACAKGE.jl.

**Comment by DilumAluthge** [comment]
This may go without saying but: We should also check every pull request to make sure that it only modifies files related to the package it claims to be related to.

For example, a pull request to register a new package `Foo.jl` would be allowed to create/modify files in the `F/Foo` directory, and would be allowed to add a line to `Registry.toml` containing the name, UUID, and path of `Foo.jl`, but would not be allowed to e.g. create/modify files in the `B/Bar` directory or modify lines in `Registry.toml` relating to the `Bar.jl` or `Baz.jl` packages. A pull request to register a new version for an existing package `Foo.jl` would be allowed to create/modify files in the `F/Foo` directory, but would not be allowed to modify `Registry.toml` or create/modify any other files.

Otherwise, we would have a security vulnerability:
- Suppose I maintain an existing package `Foo.jl`. The current version is `v1.0.0`.
- I submit a pull request to register a new version `v2.0.0` of my package `Foo.jl`. So the "sequential version number" requirement is met ✅.
- I make sure that I have `compat` entries for all dependencies including Julia, and each `compat` entry has an upper bound ✅.
- Version `v2.0.0` of `Foo.jl` is installable ✅ and loadable ✅ .
- Oh but also, by the way, my pull request changes the URL of `OrderedCollections.jl` to point to my fork of `OrderedCollections.jl` into which I have inserted some malicious code.

**Comment by StefanKarpinski** [comment]
We would only automerge PRs made by Registrator, which we control, not PRs made by people—those must be manually reviewed.

**Comment by DilumAluthge** [comment]
> We would only automerge PRs made by Registrator, which we control, not PRs made by people—those must be manually reviewed.

Perfect!

**Comment by DilumAluthge** [comment]
I have opened a pull request that implements these automatic merging guidelines: https://github.com/JuliaRegistries/General/pull/3342

**Comment by oxinabox** [comment]
> Compat for all dependencies
>   - all [deps] should also have [compat] entries (and Julia itself)
>   - [compat] entries should have upper bounds

I would like to suggest ammending this to:

> Compat for all changed/added dependencies
>   - all new [deps] should also have [compat] entries (and Julia itself)
>   - modified [compat] entries should have upper bounds

Basically, if you didn't touch a dep or its compat, then you should still get auto-merge.
I am a big fan of caret bounding everything,
but exceptions can apply.

For example lets say my package depends on DataFrames.jl,
but the only thing I actually depend upon is that DataFrames exports the `AbstractDataFrame` type, because I have some type-constraint related to that.
E.g. I just define a trait `obsidm(::Type{AbstractDataFrame})=2)`.

So I only lower-bound it, because I am pretty confidant that DataFrames is not going to stop exporting that type.
Then when I tag a release for the first time, it is blocked from automerge because of this,
and I explain to the General registry maintainer why I am only lower-bounding it.
And they are like _"Yeah, makes sense. **[merge]**"_.

Now later I update my package, I haven't changed how I use DataFrames,
I don't want to be blocked and having to reexplain each time why.

But If later I do bump my minimum DataFrames compat, then I should be stopped since my past evidence has been proven wrong, and I do need the upper-bounds.
*and* also because I need to go an ret-con the registry to say that all my old version also should be upper bounded.
